metrics/cloudwatch/cloudwatch.go

```diff
@@ -133,11 +144,41 @@ func (cw *CloudWatch) Send() error {
 		}
 	}
 
-	_, err := cw.svc.PutMetricData(&cloudwatch.PutMetricDataInput{
-		Namespace:  aws.String(cw.namespace),
-		MetricData: datums,
-	})
-	return err
+	var tokens = make(chan struct{}, cw.numConcurrentRequests)
+	var errors = make(chan error)
+	var n int
+
+	for len(datums) > 0 {
+		var batch []*cloudwatch.MetricDatum
+		lim := min(len(datums), cw.numConcurrentRequests)
+		batch, datums = datums[:lim], datums[lim:]
+		n++
+		go func(batch []*cloudwatch.MetricDatum) {
+			tokens <- struct{}{}
+			_, err := cw.svc.PutMetricData(&cloudwatch.PutMetricDataInput{
+				Namespace:  aws.String(cw.namespace),
+				MetricData: batch,
+			})
+			<-tokens
+			errors <- err
+		}(batch)
+	}
+
+	var firstErr error
+	for ; n > 0; n-- {
+		if err := <-errors; err != nil && firstErr != nil {
+			firstErr = err
+		}
+	}
```

Comment:
It's a little fragile, for a couple reasons:

- n is only loosely coupled to the number of goroutines you launched, easy to get out of sync w/ maintenance
- If you forget to do precisely as many reads from errors as goroutines you launched, you'll leak goroutines

Suggest precomputing batches, constructing errors as `make(chan error, len(batches))`, and only then launching the goroutines to process the batches. Finally, changing this loop to e.g.

```go
for i := 0; i < cap(errors); i++ {
    if err := <-errors; err != nil ... { ... }
}
```